File: 202-happy-number/happy-number.java
class Solution {
    public boolean isHappy(int n) {
        int slow = n;
        int fast = n;
        do{
            slow = helper(slow);
            fast = helper(helper(fast));
        }while(slow != fast);
        return slow == 1;
    }

    private int helper(int n){
        int res = 0;
        while(n > 0){
            int ld = n % 10;
            res += ld * ld;
            n = n / 10;
        }
        return res;
    }
}

File: 202-happy-number/happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        slow = n
        fast = n
        while(True):
            slow = self.nextL(slow)
            fast = self.nextL(self.nextL(fast))
            if slow == fast:
                break
        return slow == 1
    
    def nextL(self,n):
        sum = 0
        while n > 0:
            ld = n % 10
            sum += ld * ld
            n //= 10
        return sum